### PR TITLE
Add support for serving files directly from object storage via presigned URL redirects

### PR DESCRIPTION
**Pull request by stv0g**
This reduces the load in the Django app.

Tested on a Ceph RadosGW cluster running via rook.io in Kubernetes.

### CODE REVIEW

**Comment by konstin** [comment]
You shouldn't use the django file serving backend in prod at all, but configure a redirect from nginx to minio as it is shown in `etc/nginx.conf`

**Comment by konstin** [comment]
Unfortunately I have no clue about kubernetes, so I can't say anything about that. My understanding was that minio is a "kubernetes native" object storage, so it would be nice to integrate into a cluster (and possible to connect with disk/backblaze/s3)

**Comment by konstin** [comment]
BTW have you seen https://github.com/meine-stadt-transparent/meine-stadt-transparent/tree/main/etc/dev? It's what I for development and this already uses minio through nginx instead of django/python; I've mostly kept the python version around for making it possible to develop or run tests without an nginx running

**Comment by stv0g** [comment]
Thanks for the hints :)

**Comment by stv0g** [comment]
I've think I will keep serving redirects (presigned URLs) via Django.
This has mainly two advantages for us:

1. I dont need allow anonymous access to the S3 bucket. The presigned URLs used in the redirect are only valid for a couple of minutes
2. I can include custom HTTP headers in the presigned URL. As a result S3 delivers the files with proper Content-type and Content-disposition header which allows me to serve them with their original filename when downloaded.

I also think that the overhead is neglegible as the actual file is still served directly from S3 rather then passing through Gunicorn.

My question would be: would you consider to merge this in MST or should we keep it in our fork for the long term?

File: mainapp/views/views.py [hunk 1]
```diff
@@ -243,18 +247,51 @@ def file_serve_proxy(
...
-
     """ Ensure that the file is not deleted in the database """
-    get_object_or_404(File, id=id)
+    file = get_object_or_404(File, id=id)
 
-    minio_file = minio_client().get_object(minio_file_bucket, str(id))
-    response = HttpResponse(minio_file.read())
+    name, ext = splitext(file.filename)
+    if name.isnumeric() and file.name:
+        filename = f'{file.name}_{name}{ext}'
```
Comment (konstin) [inline mainapp/views/views.py:255]: is file.name save to use here, both regarding used characters and length?

File: mainapp/views/views.py [hunk 2]
```diff
@@ -243,18 +247,51 @@ def file_serve_proxy(
...
+        if settings.MINIO_PUBLIC_HOST:
+            endpoint = settings.MINIO_PUBLIC_HOST
+            secure = settings.MINIO_PUBLIC_SECURE
+        else:
+            endpoint = settings.MINIO_HOST
+            secure = settings.MINIO_SECURE
+
+        mc = minio_client()
+        old_base_url = mc._base_url
+        mc._base_url = minio.api.BaseURL(("https://" if secure else "http://") + endpoint, settings.MINIO_REGION)
```
Comment (konstin) [inline mainapp/views/views.py:277]: please avoid using private variables. in this case you can pass the settings on the `minio_client` function or just construct a new minio client object in the spot

**Review by konstin — APPROVED** [review]
i didn't realize at first that this is actually using a redirect and that python needs to be involved to build the URL, so this is a nice addition!

File: mainapp/views/views.py (hunk 1 above)
Comment (stv0g) [reply mainapp/views/views.py:255]: Seems like this is not an easy topic.

There is [RFC5987](https://www.rfc-editor.org/rfc/rfc5987) which defines the encoding of UTF-8 strings in Header fields.

Also it seems like the filename needs to be percentage-encoded.

See also: https://stackoverflow.com/questions/93551/how-to-encode-the-filename-parameter-of-content-disposition-header-in-http

I will push a fix in a minute.

File: mainapp/views/views.py (hunk 2 above)
Comment (stv0g) [reply mainapp/views/views.py]: Agreed. That was a nasty workaround.

I fixed it by creating two Minio instances of which one uses the public S3 endpoint of configured with MINIO_PUBLIC_HOST. This instance is only used for generating the presigned URLs. All the actual S3 operations are targeting always the default endpoint.

**Comment by stv0g** [comment]
Hi @konstin,

thanks for merging the other PRs :)

I have now rebased this PR on the latest state of `main`, addressed your comments and fixed the CI tests.

Cheers,
Steffen

File: mainapp/views/views.py [hunk 3]
```diff
@@ -242,18 +244,45 @@ def file_serve_proxy(
...
+        filename = file.filename
 
-    minio_file = minio_client().get_object(minio_file_bucket, str(id))
-    response = HttpResponse(minio_file.read())
+    filename_safe = filename.encode("ascii", "ignore")
 
-    response["Content-Type"] = minio_file.headers["Content-Type"]
+    headers = {
+        # Encoding according to RFC5987
+        "Content-Disposition": f"attachment; filename=\"{quote(filename_safe)}\"; filename*=UTF-8''{quote(filename)}",
```
Comment (konstin) [inline mainapp/views/views.py:260]: That's smarter than what I'd have used :+1:

**Comment by konstin** [comment]
Thank you!

**Pull request merged**